### ISSUE TITLE
Schema-first, Resolver inheritance not working after upgrade

### ISSUE THREAD

**Issue by isbobbydigitalak** [graphql-dotnet/graphql-dotnet#3194]
Hello!

I am currently upgrading from a very old version of the library (2.4.0) to the 5.3.0 version. Of course a lot of things have changed since then and I have been referring to the migration guides to find explanations for various places that broke and how to fix them, which has been really good.

One place however that I haven't been able to figure out is why resolver types don't seem to inherit the resolver methods from parent types anymore or if there is a way to get that working. We are using schema first so currently we're simply loading in the schema text with Schema.For and in the configuration lambda we're calling `_.Types.Include(Type)` with types that look like this:

```csharp
    [GraphQLMetadata("User", IsTypeOf = typeof(User))]
    public class UserType : IAPIType
    {
        public Guid Id([FromSource] User user) => user.Id;
    /// ...
     }
``` 

These cases are now working fine but for some tpes where we had a base type that implements resolvers common to all the child types it is no longer working. 

The parent type looks like this 

```csharp
    public class NodeType<T> : IAPIType where T : Node
    {
        protected readonly IServiceProvider _serviceProvider;

        public NodeType(IServiceProvider serviceProvider)
        {
            _serviceProvider = serviceProvider;
        }

        /// one of the parent resolver methods that we want to inherit
        public Guid Id([FromSource] T node) => node.Id;

        public string Identifier([FromSource] T node) => node.Identifier;
        // ....
     }
```

And one if its children is:

```csharp
    [GraphQLMetadata("FolderNode", IsTypeOf = typeof(FolderNode))]
    public class FolderNodeType : NodeType<FolderNode>
    {
        public FolderNodeType(IServiceProvider serviceProvider) : base(serviceProvider)
        {
        }

        public new Guid? ParentId([FromSource] FolderNode node) => node.Id;
    } 
```

However when trying to resolve the Id of a FolderNode (attempting to rely  on the implementation received from NodeType) it gives the following error:

```
Exception thrown: 'System.InvalidOperationException' in Api.dll: 'Could not resolve an instance of NodeType`1 to execute FolderNode.id'
```

I can't really figure out what's going wrong here based on the notes in the migration guides. And it seems like since it's inheriting from the NodeType parent it should be getting its Id method for free. Would be a shame to have to replicate these methods over all the child types..

Got any ideas of what I'm missing on this one?

Thanks!

**Comment by Shane32** [comment]
Within `AutoRegisteringObjectGraphType<TSourceType>.GetRegisteredMembers`, the call to `GetMethods` includes the `BindingFlags.DeclaredOnly` flag, which will exclude methods of inherited members.  Probably this is the cause.  Unfortunately I don't know schema-first enough to immediately know how it's tied to that class.  If it's tied in via DI, then you can easily inherit the class and override the method, and register your custom class over top of the regular one.  Perhaps we can also issue a fix/change for this in v7.

At the moment I would try this to see if it works:

```cs

    public class MyAutoRegisteringObjectGraphType<TSourceType> : AutoRegisteringObjectGraphType<TSourceType>
    {
        protected override IEnumerable<MemberInfo> GetRegisteredMembers()
        {
            var props = typeof(TSourceType).GetProperties(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static).Where(x => x.CanRead);
            var methods = typeof(TSourceType).GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
                .Where(x =>
                    !x.ContainsGenericParameters &&               // exclude methods with open generics
                    !x.IsSpecialName &&                           // exclude methods generated for properties
                    x.ReturnType != typeof(void) &&               // exclude methods which do not return a value
                    x.ReturnType != typeof(Task) &&               // exclude methods which do not return a value
                    //x.GetBaseDefinition() == x &&                 // exclude methods which override an inherited class' method (e.g. GetHashCode)
                                                                  // exclude methods generated for record types: bool Equals(TSourceType)
                    !(x.Name == "Equals" && !x.IsStatic && x.GetParameters().Length == 1 && x.GetParameters()[0].ParameterType == typeof(TSourceType) && x.ReturnType == typeof(bool)) &&
                    x.Name != "<Clone>$");                        // exclude methods generated for record types
            return props.Concat<MemberInfo>(methods);
        }
    }

// map AutoRegisteringObjectGraphType<> to MyAutoRegisteringObjectGraphType<>
services.AddTransient(typeof(AutoRegisteringObjectGraphType<>), typeof(MyAutoRegisteringObjectGraphType<>));
```

It is likely that the above patch might cause `MemberwiseClone()` or `GetHashCode()` or some other methods to be included, and you will need to add a little more code to exclude those methods.

If it doesn't work, I'll need to investigate the schema builder to try to figure out how these classes are tied together.  Of course I could be wrong, and the issue is somewhere else, but I feel it's likely.

**Comment by Shane32** [comment]
I guess I'll have to reproduce the problem to diagnose it further.

**Comment by Shane32** [comment]
I found the problem.  It's a bug, easily fixed in the source, but without an easy way to patch on your end.

**Comment by Shane32** [comment]
Fixed in version 5.3.1, just released.

**Comment by isbobbydigitalak** [comment]
Wow, that was fast! Very happy I decided to open an issue instead of banging my head against a wall since it turned out to be a bug! 

Thanks a lot!
